### PR TITLE
DOC: Cite code base references using `sphinxcontrib-bibtex`

### PR DESCRIPTION
**Pull request by jhlegarreta**
- DOC: Miscellaneous improvements to references file
- DOC: Cite code base references using `sphinxcontrib-bibtex`

### CODE REVIEW

**Comment by jhlegarreta** [comment]
I have been unable to find the citation information for this:
https://github.com/dipy/dipy/blob/1cc364f2b20479b29063dd7ed231d9ab3fcd710f/dipy/reconst/gqi.py#L173

@Garyfallidis would you please provide me with the necessary information? As the current PR touches a lot of files, and to expedite merging and avoid conflicts, it can be left for a separate PR.

**Comment by jhlegarreta** [comment]
So it looks like cross-referencing in the short description is not something that `sphinxcontrib-bibtex` enjoys, even for functions. I am not very fond of adding a separate, verbose paragraph for all refs saying ` See :footcite:`author` for further details.`. @skoudoro let me know if you discover what's missing to make these work without needing it.

**Comment by skoudoro** [comment]
> let me know if you discover what's missing to make these work without needing it. Thanks.

We need to activate `numpydoc` and it should improve cross-referencing in the short description. and in general.

Can you uncomment https://github.com/dipy/dipy/blob/5e074ddf9564f5298bacb2d1611ad79d8d3077f5/doc/conf.py#L43.

if build fails when you uncomment, we need to figure out this first, then it should improve your issue when citation. 

keep me updated and thanks for the HUGE work ! 

I will review later today or tomorrow

**Comment by jhlegarreta** [comment]
> We need to activate numpydoc and it should improve cross-referencing in the short description. and in general.

Done.

**Comment by skoudoro** [comment]
Maybe @larsoner will have an idea on how to fix this kind of issue (`ERROR: Unknown target name: "footcite-fieremans2011".`).   here: https://github.com/dipy/dipy/actions/runs/10422480080/job/28867046186?pr=3316#step:5:1155

I am not sure why it is not recognized and we are doing wrong. It seems to wrong correctly on mne-python.

File: dipy/reconst/dki_micro.py [hunk 1]
```diff
@@ -320,7 +324,8 @@ class KurtosisMicrostructureModel(DiffusionKurtosisModel):
     """Class for the Diffusion Kurtosis Microstructural Model"""
 
     def __init__(self, gtab, fit_method="WLS", *args, **kwargs):
-        """Initialize a KurtosisMicrostrutureModel class instance [1]_.
+        """Initialize a KurtosisMicrostrutureModel class instance
+        :footcite:p:`Fieremans2011`_.
```
Comment (larsoner) [inline dipy/reconst/dki_micro.py:328]: This is probably an error
```suggestion
        :footcite:p:`Fieremans2011`.
```

File: dipy/reconst/ivim.py [hunk 2]
```diff
@@ -513,7 +509,8 @@ def _leastsq(self, data, x0):
 
 class IvimModelVP(ReconstModel):
     def __init__(self, gtab, bounds=None, maxiter=10, xtol=1e-8):
-        r"""Initialize an IvimModelVP class.
+        r"""Initialize an IvimModelVP class :footcite:p:`LeBihan1988`,
+        :footcite:p:`Federau2012`, :footcite:p:`Fadnavis2019`.
```
Comment (larsoner) [inline dipy/reconst/ivim.py:513]: I have never tried putting footcite in a title, maybe it's problematic for some reason?

**Review by larsoner — COMMENTED** [review]
Not sure why there are so many errors, it looks like you're using things correctly. Might be worth looking at the built HTML to see what those errant links actually end up looking like

**Comment by skoudoro** [comment]
Thank you for the feedback, we will look deeper

File: dipy/reconst/dki_micro.py (hunk 1 above)
Comment (jhlegarreta) [reply dipy/reconst/dki_micro.py]: :+1: Fixed in 8412ffb.

File: dipy/reconst/ivim.py (hunk 2 above)
Comment (jhlegarreta) [reply dipy/reconst/ivim.py:513]: Although this was assumed to be the case here:
https://github.com/dipy/dipy/pull/3310#issuecomment-2287664588

When building locally, this is rendered correctly:
https://github.com/dipy/dipy/pull/3316/files#diff-ba7f8dd2f3a9bd5b0c4da601efc4abc84940839ab91c88e0742ee216399ea9f1R299

So not sure what the problem is.

**Comment by jhlegarreta** [comment]
I have tried only modifying the refs in the `data.fetcher.read_qtdMRI_test_retest_2subjects` method, leaving the rest of the code base untouched, and the refs are not parsed correctly; those two refs have nothing special either, so not sure what we are facing here after our successful changed in #2810 and #3310.

**Comment by jhlegarreta** [comment]
Re https://github.com/dipy/dipy/pull/3316#issuecomment-2294890031 did some more investigation, and despite the `ParticleFilteringTracking` case, the short description problem looks like the most plausible hypothesis here:
- For the `ParticleFilteringTracking` case, the class does not have a short description; its `__init__` has. Thus, the `tracking` module HTML is built correctly, and the reference is processed correctly.
- For a local test modifying the `data.fetcher.read_qtdMRI_test_retest_2subjects` method and adopting the `footcite:p:{}` convention (and leaving the rest of the code base untouched), the `data` module HTML layout is broken, and refs error.

The only difference I can see between the two is that `read_qtdMRI_test_retest_2subjects` is shown in a table (generated following the toctree I guess), whereas for `ParticleFilteringTracking`, as the class does not have a short description, its corresponding short description entry in the table is empty, and thus there is nothing to be broken due to the  `numpydoc` short description duplication issue https://github.com/dipy/dipy/pull/3310#issuecomment-2287707032, and as its `__init__` method is not shown in any table whatsoever, the module documentation is rendered correctly, including the reference.

If I put the  `data.fetcher.read_qtdMRI_test_retest_2subjects` method references in a separate paragraph, the `data` module HTML is not broken, the refs are parsed correctly, and the number of documentation errors/warnings is also reduced wrt the case where the refs dwell in the short description.

So despite https://github.com/dipy/dipy/pull/3316#issuecomment-2292460935, it looks like we do not have any other choice but to reword all short descriptions affected by the refs.

@skoudoro Let me know what your local investigation gives.

**Comment by jhlegarreta** [comment]
Reported failures are unrelated to this patch set: we are getting HTTP errors for some of the files that the fetcher requires to download.

**Comment by skoudoro** [comment]
So, I spend time on it and I came to the same conclusion as you: the short description can not have the citation.

Also, I looked at what we had before, and it was not working correctly in the short description. It seems it is not specific to `sphinxcontrib-bibtex`. it seems the toctree summary is confused by those references. So it disable them. 

Conclusion:  **YES this a rule that we will need to enforce**,  starting from this PR. 

>  I am not very fond of adding a separate, verbose paragraph for all refs saying

Let's say it is an opportunity to improve some short description (by making them imperative and very short) and provide value in the long description. 

Let me know if you need help to update this PR. I believe everything will be back to normal after updating that.

thank you @jhlegarreta

**Comment by jhlegarreta** [comment]
@skoudoro OK. Have not added the imperative style, and have only modified those that had a citation; these changes have already required a non-negligible amount of effort and time, so let's leave the imperative style for when we will be enforcing it through the appropriate `numpydoc` flag, as well as making the shorts descriptions short for the docstrings.

So it is ready to be merged.

HTTP errors have been recurring since a few days when fetching the data, so unrelated build failures should be expected; the doc build should come green with all code base ref warnings solved.

File: dipy/segment/clustering_algorithms.pyx [hunk 3]
```diff
@@ -66,7 +66,7 @@ def peek(iterable):
 @warning_for_keywords()
 def quickbundles(streamlines, Metric metric, double threshold, *,
                  long max_nb_clusters=BIGGEST_INT, ordering=None):
-    """ Clusters streamlines using QuickBundles.
+    """ Clusters streamlines using QuickBundles :footcite:p:`Garyfallidis2012a`.
```
Comment (skoudoro) [inline dipy/segment/clustering_algorithms.pyx:69]: To move out from short description

File: dipy/segment/clustering_algorithms.pyx [hunk 4]
```diff
@@ -122,7 +120,8 @@ def quickbundles(streamlines, Metric metric, double threshold, *,
 
 @warning_for_keywords()
 def quickbundlesx(streamlines, Metric metric, thresholds, *, ordering=None):
-    """ Clusters streamlines using QuickBundlesX.
+    """ Clusters streamlines using QuickBundlesX
+    :footcite:p:`Garyfallidis2012a`, :footcite:p:`Garyfallidis2016`.
```
Comment (skoudoro) [inline dipy/segment/clustering_algorithms.pyx:124]: To move out from short description

File: dipy/segment/fss.py [hunk 5]
```diff
@@ -26,7 +26,7 @@ def __init__(
 
         Generate the Binned K-D Tree structure with reference streamlines,
         using streamlines barycenter and mean-points.
-        See [StOnge2022]_ for further details.
+        See :footcite:p:`StOnge2022` for further details.
```
Comment (skoudoro) [inline dipy/segment/fss.py:29]: To move out from short description

File: dipy/tracking/local_tracking.py [hunk 6]
```diff
@@ -296,7 +296,7 @@ def __init__(
         initial_directions=None,
     ):
         r"""A streamline generator using the particle filtering tractography
-        method [1]_.
+        method :footcite:p:`Girard2014`.
```
Comment (skoudoro) [inline dipy/tracking/local_tracking.py:299]: To move out from short description

File: dipy/tracking/stopping_criterion.pyx [hunk 7]
```diff
@@ -166,7 +166,9 @@ cdef class AnatomicalStoppingCriterion(StoppingCriterion):
 
 cdef class ActStoppingCriterion(AnatomicalStoppingCriterion):
     r"""
-    Anatomically-Constrained Tractography (ACT) stopping criterion from [1]_.
+    Anatomically-Constrained Tractography (ACT) stopping criterion from
+    :footcite:p:`Smith2012`.
```
Comment (skoudoro) [inline dipy/tracking/stopping_criterion.pyx:170]: To move out from short description

File: dipy/tracking/stopping_criterion.pyx [hunk 8]
```diff
@@ -226,7 +225,9 @@ cdef class ActStoppingCriterion(AnatomicalStoppingCriterion):
 
 cdef class CmcStoppingCriterion(AnatomicalStoppingCriterion):
     r"""
-    Continuous map criterion (CMC) stopping criterion from [1]_.
+    Continuous map criterion (CMC) stopping criterion from
+    :footcite:p:`Girard2014`.
+
```
Comment (skoudoro) [inline dipy/tracking/stopping_criterion.pyx:230]: To move out from short description

File: dipy/tracking/streamlinespeed.pyx [hunk 9]
```diff
@@ -541,22 +541,23 @@ cdef cnp.npy_intp c_compress_streamline(Streamline streamline, Streamline out,
 
 @warning_for_keywords()
 def compress_streamlines(streamlines, *, tol_error=0.01, max_segment_length=10):
-    """ Compress streamlines by linearization as in [Presseau15]_.
+    """ Compress streamlines by linearization as in :footcite:p:`Presseau2015`.
```
Comment (skoudoro) [inline dipy/tracking/streamlinespeed.pyx:544]: To move out from short description

**Review by skoudoro — COMMENTED** [review]
Thank you for the quick update @jhlegarreta.

See below citations to move out from short description. After that, it should be good to go. if we are missing more it will be in a new PR. 

Thanks!

**Comment by skoudoro** [comment]
Also, Exciting to see this:

image

it's coming zero warning policy!

**Comment by jhlegarreta** [comment]
All the comments above relate to either private methods, methods that do not appear on tables, or Cython methods that are not listed, so I consciously did not change the short description vs long description. But will do now. Last change hopefully.

File: dipy/reconst/ivim.py [hunk 10]
```diff
@@ -561,7 +549,7 @@ def __init__(self, gtab, bounds=None, maxiter=10, xtol=1e-8):
     def fit(self, data, bounds_de=None, **kwargs):
         r"""Fit method of the IvimModelVP model class
 
-        MicroLearn framework (VarPro)[1]_.
+        MicroLearn framework (VarPro) :footcite:p:`Fadnavis2019`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/ivim.py:552]: This is already the long description.

File: dipy/segment/fss.py (hunk 5 above)
Comment (jhlegarreta) [reply dipy/segment/fss.py:29]: This is already the long description.

File: dipy/align/streamlinear.py [hunk 11]
```diff
@@ -315,7 +315,8 @@ def __init__(
         evolution=False,
         num_threads=None,
     ):
-        r"""Linear registration of 2 sets of streamlines [Garyfallidis15]_.
+        r"""Linear registration of 2 sets of streamlines
+        :footcite:p:`Garyfallidis2015`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/align/streamlinear.py]: Done in ceefbc0.

File: dipy/core/optimize.py [hunk 12]
```diff
@@ -386,7 +386,8 @@ def fit(self, X, y):
 class PositiveDefiniteLeastSquares:
     @warning_for_keywords()
     def __init__(self, m, *, A=None, L=None):
-        r"""Regularized least squares with linear matrix inequality constraints [1]_.
+        r"""Regularized least squares with linear matrix inequality constraints
+        :footcite:p:`DelaHaije2020`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/core/optimize.py]: Done in ceefbc0.

File: dipy/denoise/gibbs.py [hunk 13]
```diff
@@ -145,7 +145,8 @@ def _gibbs_removal_1d(x, *, axis=0, n_points=3):
 
 def _weights(shape):
     """Computes the weights necessary to combine two images processed by
-    the 1D Gibbs removal procedure along two different axes [1]_.
+    the 1D Gibbs removal procedure along two different axes
+    :footcite:p:`Kellner2016`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/denoise/gibbs.py]: Done in ceefbc0.

File: dipy/denoise/nlmeans.py [hunk 14]
```diff
@@ -22,7 +22,8 @@ def nlmeans(
     rician=True,
     num_threads=None,
 ):
-    r"""Non-local means for denoising 3D and 4D images [Descoteaux08]_.
+    r"""Non-local means for denoising 3D and 4D images
+    :footcite:p:Descoteaux2008a`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/denoise/nlmeans.py]: Done in ceefbc0.

File: dipy/denoise/nlmeans_block.pyx [hunk 15]
```diff
@@ -348,6 +348,7 @@ cpdef upfir(double[:, :] image, double[:] h):
 @cython.cdivision(True)
 def nlmeans_block(double[:, :, :]image, double[:, :, :] mask, int patch_radius, int block_radius, double h, int rician):
     """Non-Local Means Denoising Using Blockwise Averaging
+    :footcite:p:`Coupe2011`, :footcite:p:`Coupe2012`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/denoise/nlmeans_block.pyx]: Done in ceefbc0.

File: dipy/denoise/shift_twist_convolution.pyx [hunk 16]
```diff
@@ -17,7 +17,9 @@ from dipy.testing.decorators import warning_for_keywords
 @warning_for_keywords()
 def convolve(odfs_sh, kernel, sh_order_max, *, test_mode=False, num_threads=None, normalize=True):
     """ Perform the shift-twist convolution with the ODF data and
-    the lookup-table of the kernel.
+    the lookup-table of the kernel :footcite:p:`Meesters2016`,
+    :footcite:p:`Duits2011`, :footcite:p:`Portegies2015a`,
+    :footcite:p:`Portegies2015b`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/denoise/shift_twist_convolution.pyx]: Done in ceefbc0.

File: dipy/reconst/cti.py [hunk 17]
```diff
@@ -199,7 +199,7 @@ class CorrelationTensorModel(ReconstModel):
     """Class for the Correlation Tensor Model"""
 
     def __init__(self, gtab1, gtab2, fit_method="WLS", *args, **kwargs):
-        """Correlation Tensor Imaging Model [1]_
+        """Correlation Tensor Imaging Model :footcite:p:`NetoHenriques2020`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/cti.py]: Done in ceefbc0.

File: dipy/reconst/csdeconv.py [hunk 18]
```diff
@@ -195,19 +193,23 @@ def __init__(
         tau=0.1,
         convergence=50,
     ):
-        r"""Constrained Spherical Deconvolution (CSD) [1]_.
+        r"""Constrained Spherical Deconvolution (CSD)
+        :footcite:p:`Tournier2007`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/csdeconv.py]: Done in ceefbc0.

File: dipy/reconst/dti.py [hunk 19]
```diff
@@ -698,7 +684,8 @@ class TensorModel(ReconstModel):
     """Diffusion Tensor"""
 
     def __init__(self, gtab, fit_method="WLS", return_S0_hat=False, *args, **kwargs):
-        """A Diffusion Tensor Model [1]_, [2]_.
+        """A Diffusion Tensor Model :footcite:p:`Basser1994b`,
+        :footcite:p:`Basser1996`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/dti.py]: Done in ceefbc0.

File: dipy/reconst/dki_micro.py [hunk 20]
```diff
@@ -320,7 +327,8 @@ class KurtosisMicrostructureModel(DiffusionKurtosisModel):
     """Class for the Diffusion Kurtosis Microstructural Model"""
 
     def __init__(self, gtab, fit_method="WLS", *args, **kwargs):
-        """Initialize a KurtosisMicrostrutureModel class instance [1]_.
+        """Initialize a KurtosisMicrostrutureModel class instance
+        :footcite:p:`Fieremans2011`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/dki_micro.py]: Done in ceefbc0.

File: dipy/reconst/forecast.py [hunk 21]
```diff
@@ -58,9 +51,12 @@ def __init__(
         lambda_csd=1.0,
     ):
         r"""Analytical and continuous modeling of the diffusion signal with
-        respect to the FORECAST basis [1,2,3]_.
+        respect to the FORECAST basis :footcite:p:`Anderson2005`,
+        :footcite:p:`Kaden2016a`, :footcite:p:`Zucchelli2017`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/forecast.py]: Done in ceefbc0.

File: dipy/reconst/fwdti.py [hunk 22]
```diff
@@ -90,7 +87,7 @@ class FreeWaterTensorModel(ReconstModel):
     """Class for the Free Water Elimination Diffusion Tensor Model"""
 
     def __init__(self, gtab, fit_method="NLS", *args, **kwargs):
-        """Free Water Diffusion Tensor Model [1]_.
+        """Free Water Diffusion Tensor Model :footcite:p:`NetoHenriques2017`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/fwdti.py]: Done in ceefbc0.

File: dipy/reconst/gqi.py [hunk 23]
```diff
@@ -11,14 +11,14 @@
 
 class GeneralizedQSamplingModel(OdfModel, Cache):
     def __init__(self, gtab, method="gqi2", sampling_length=1.2, normalize_peaks=False):
-        r"""Generalized Q-Sampling Imaging [1]_
+        r"""Generalized Q-Sampling Imaging :footcite:t:`Yeh2010`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/gqi.py]: Done in ceefbc0.

File: dipy/reconst/ivim.py [hunk 24]
```diff
@@ -186,7 +186,8 @@ def __init__(
         maxiter=1000,
     ):
         r"""
-        Initialize an IVIM model.
+        Initialize an IVIM model :footcite:p:`LeBihan1988`,
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/ivim.py]: Done in ceefbc0.

File: dipy/reconst/mapmri.py [hunk 25]
```diff
@@ -24,57 +24,33 @@
 
 
 class MapmriModel(ReconstModel, Cache):
-    r"""Mean Apparent Propagator MRI (MAPMRI) [1]_ of the diffusion signal.
+    r"""Mean Apparent Propagator MRI (MAPMRI) footcite:p:`Ozarslan2013` of the
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/mapmri.py]: Done in ceefbc0.

File: dipy/reconst/msdki.py [hunk 26]
```diff
@@ -274,7 +273,7 @@ class MeanDiffusionKurtosisModel(ReconstModel):
     """Mean signal Diffusion Kurtosis Model"""
 
     def __init__(self, gtab, bmag=None, return_S0_hat=False, *args, **kwargs):
-        """Mean Signal Diffusion Kurtosis Model [1]_.
+        """Mean Signal Diffusion Kurtosis Model :footcite:p:`NetoHenriques2018`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/msdki.py]: Done in ceefbc0.

File: dipy/reconst/rumba.py [hunk 27]
```diff
@@ -38,7 +38,8 @@ def __init__(
         verbose=False,
     ):
         """
-        Robust and Unbiased Model-BAsed Spherical Deconvolution (RUMBA-SD) [1]_
+        Robust and Unbiased Model-BAsed Spherical Deconvolution (RUMBA-SD)
+        :footcite:p:`CanalesRodriguez2015`.
```
[the remark this answers is not in the document]
Comment (jhlegarreta) [reply dipy/reconst/rumba.py]: Done in ceefbc0.

File: dipy/segment/clustering_algorithms.pyx (hunk 3 above)
Comment (jhlegarreta) [reply dipy/segment/clustering_algorithms.pyx]: Done in ceefbc0.

File: dipy/segment/clustering_algorithms.pyx (hunk 4 above)
Comment (jhlegarreta) [reply dipy/segment/clustering_algorithms.pyx]: Done in ceefbc0.

File: dipy/tracking/local_tracking.py (hunk 6 above)
Comment (jhlegarreta) [reply dipy/tracking/local_tracking.py]: Done in ceefbc0.

File: dipy/tracking/stopping_criterion.pyx (hunk 7 above)
Comment (jhlegarreta) [reply dipy/tracking/stopping_criterion.pyx]: Done in ceefbc0.

File: dipy/tracking/stopping_criterion.pyx (hunk 8 above)
Comment (jhlegarreta) [reply dipy/tracking/stopping_criterion.pyx]: Done in ceefbc0.

File: dipy/tracking/streamlinespeed.pyx (hunk 9 above)
Comment (jhlegarreta) [reply dipy/tracking/streamlinespeed.pyx]: Done in ceefbc0.

**Review by skoudoro — APPROVED** [review]
Thank you for this huge work @jhlegarreta,

Merging.

**Pull request merged**